pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,391 @@
+	oldTsDir, err := os.Readlink(dataDirPath)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("%s: error reading link for data directory: %v", w.LogContext, err)
+		return err
+	}
+
+	newDataDirPath := path.Join(w.TargetDir, newDataDirName)
+	if err = os.Symlink(tsDirName, newDataDirPath); err != nil {
+		os.RemoveAll(tsDir)
+		glog.Errorf("%s: error creating symbolic link for atomic update: %v", w.LogContext, err)
+		return err
+	}
+
+	if err = os.Rename(newDataDirPath, dataDirPath); err != nil {
+		os.Remove(newDataDirPath)
+		os.RemoveAll(tsDir)
+		glog.Errorf("%s: error renaming symbolic link for data directory %s: %v", w.LogContext, newDataDirPath, err)
+		return err
+	}
+
+	if len(oldTsDir) > 0 {
+		if err = os.RemoveAll(path.Join(w.TargetDir, oldTsDir)); err != nil {
+			glog.Errorf("%s: error removing old data directory %s: %v", w.LogContext, oldTsDir, err)
+			return err
+		}
+	}
+
+	if err = w.removeUserVisiblePaths(pathsToRemove); err != nil {
+		glog.Errorf("%s: error pruning old visible symlinks: %v", w.LogContext, err)
+		return err
+	}
+
+	return w.touchSentinelFile()
+}
+
+// validatePayload returns an error if any path in the payload  returns a copy of the payload with the paths cleaned.
+func validatePayload(payload map[string][]byte) (map[string][]byte, error) {
+	cleanPayload := make(map[string][]byte)
+	for k, content := range payload {
+		if strings.Contains(k, "..") {
+			return nil, fmt.Errorf("invalid path: %q; paths must not contain '..'", k)
+		}
+
+		cleanPayload[path.Clean(k)] = content
+	}
+
+	return cleanPayload, nil
+}
+
+// shouldWritePayload returns whether the payload should be written to disk.
+func (w *AtomicWriter) shouldWritePayload(payload map[string][]byte) (bool, error) {
+	for userVisiblePath, content := range payload {
+		shouldWrite, err := w.shouldWriteFile(path.Join(w.TargetDir, userVisiblePath), content)
+		if err != nil {
+			return false, err
+		}
+
+		if shouldWrite {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// shouldWriteFile returns whether a new version of a file should be written to disk.
+func (w *AtomicWriter) shouldWriteFile(path string, content []byte) (bool, error) {
+	_, err := os.Lstat(path)
+	if os.IsNotExist(err) {
+		return true, nil
+	}
+
+	contentOnFs, err := ioutil.ReadFile(path)
+	if err != nil {
+		return false, err
+	}
+
+	return (bytes.Compare(content, contentOnFs) != 0), nil
+}
+
+// pathsToRemove walks the user-visible portion of the target directory and
+// determines which paths should be pruned (if any) after the payload is
+// written to the target directory.
+func (w *AtomicWriter) pathsToRemove(payload map[string][]byte) (sets.String, error) {
+	paths := sets.NewString()
+	visitor := func(path string, info os.FileInfo, err error) error {
+		if path == w.TargetDir {
+			return nil
+		}
+
+		relativePath := strings.TrimPrefix(path, w.TargetDir)
+		relativePath = strings.TrimPrefix(relativePath, "/")
+		if strings.HasPrefix(relativePath, "..") {
```

Comment:
Thinking twice `..sentinel`, `..data` and the magic tsformatted dir are the only paths I would forbid in the payload (and obviously the ".."), since these are the only paths that `AtomicWriter` generates.